cmake/LibraryDefine.cmake

```diff
@@ -55,6 +55,8 @@ function(OPENEXR_DEFINE_LIBRARY libname)
     CXX_STANDARD_REQUIRED ON
     CXX_EXTENSIONS OFF
     POSITION_INDEPENDENT_CODE ON
+    C_VISIBILITY_PRESET hidden
```

Comment:
Setting the visibility preset to hidden is necessary for the export attributes to mean anything under gcc/clang, but I want to point out that the existing set up of exporting individual members as we did is compatible with that approach, and should work in and off itself without changing the export mechanism on all the classes.

Reply:
yes, my original goal was to just change to turn on the hidden flag (and change the export defines, similar to what USD does) to avoid symbol leakage - we've had some issues with this recently in other libraries, but seemed an appropriate opportunity to become tidier about what symbols we expose to the world.

However, once I set the default visibility to hidden, I was having issues with vtables not being exported consistently, even though the destructor was exported - I think something is/was off with default provided functions (i.e. rule of 3/5) and visibility. I do see what you mean by it exporting contained types as weak dynamic symbols, I did not expect that.

I will explore more why I was having issues with the vtables, to be able to clean that up, and maybe take a different stab at this.

Reply:
The problem is that gcc doesn't export the vtable. There's a proposal in the USD Issues which contains a lot more analysis of the horrible state of shared libraries in general. It also has me relating some of the OpenEXR experience...

https://github.com/PixarAnimationStudios/USD/issues/665


```
#if HAVE_GCC_OR_CLANG
#define FOO_EXPORT_API the visibility thing
#else
#define FOO_EXPORT_API
#endif

class FOO_EXPORT_API Foo
{
     int foo(); // not exported
     FOO_EXPORT int bar(); // exported
};
```

In general, this stuff is a world of hurt. More "fascinating" investigation here, wrt pybind11: https://github.com/pybind/pybind11/issues/2132

Reply:
After reading the USD issue and

https://gcc.gnu.org/onlinedocs/gcc/Vague-Linkage.html

I think it makes the most sense to follow the basic guidelines outlined in the USD issue, and have implemented similar constructs here. I also put a brief doxygen description of the issue in IexExports.h, and put all the other export headers into the same export macros group